cpu/stm32/Makefile.dep

```diff
@@ -79,4 +79,14 @@ ifneq (,$(filter periph_vbat,$(USEMODULE)))
   FEATURES_REQUIRED += periph_adc
 endif
 
+ifneq (,$(filter periph_fmc_%,$(USEMODULE)))
+  USEMODULE += periph_fmc
+endif
+
+ifneq (,$(filter periph_fmc,$(USEMODULE)))
+  FEATURES_OPTIONAL += periph_fmc_16bit
+  FEATURES_OPTIONAL += periph_fmc_nor_sram
+  FEATURES_OPTIONAL += periph_fmc_sdram
+endif
```

Comment:
:thinking: good question.In this case, I'm really not sure how to do it right. What I intended was that the user cannot decide which features are used, but that is decided by the defined features of the board, because FMC only works if all necessary modules are enabled. It would not work if the user only enabled `periph_fmc` for example.

I think the situation for FMC is a bit different than for `periph_uart_*` or `periph_spi_*` where additional features are enabled but the peripheral already works with the base module.

Would it be an option just to remove the circular dependency?
```suggestion
ifneq (,$(filter periph_fmc,$(USEMODULE)))
  FEATURES_OPTIONAL += periph_fmc_16bit
  FEATURES_OPTIONAL += periph_fmc_nor_sram
  FEATURES_OPTIONAL += periph_fmc_sdram
endif
```

Reply:
Alternatively we could add in board's `Makefile.dep` for SDRAM for example:
```c
ifneq (,$(filter periph_fmc,$(USEMODULE)))
  FEATURES_REQUIRED += periph_fmc_16bit
  FEATURES_REQUIRED += periph_fmc_sdram
endif
```
But this would require to add required features in `Makefile.features` and in `Makefile.dep`.

Reply:
Just do like I proposed in https://github.com/RIOT-OS/RIOT/pull/19843/files/33c8ff931ee36c0762f0d8e46fe14c7480bc3221#diff-b85e38a85615fbf0712b3a1de6006c631efbdb025ed9d8fd96d8df28f49f7656. Since `periph_fmc*` are defined as features, only the boards that provide these features can build the application. Using `FEATURES_REQUIRED += periph_fmc` in `cpu/stm32/Makefile.dep` does the magic. And it is important to use FEATURES_REQUIRED instead of USEMODULE.

Reply:
At least the features `periph_fmc_16bit` and `periph_fmc_32bit` must not be understood as optional additional functionalities, but as mandatory functionalities in case the FMC has to operate with 16 bits and 32 bits respectively.

The `periph_sdram` and `periph_nor_sram` features, on the other hand, could indeed be considered optional to control whether the SDRAM or the SRAM is used, e.g. when FMC is used for display but the SDRAM or the SRAM should not be used :thinking:

Reply:
And later when FMC will be used for display stuff, we could add:

```mk
ifneq (,$(filter disp_dev,$(USEMODULE)))
    FEATURES_REQUIRED += periph_fmc_sdram
endif
```

in a board `Makefile.dep` if it should use the sdram fmc with its display.

Reply:
> At least the features `periph_fmc_16bit` and `periph_fmc_32bit` must not be understood as optional additional functionalities, but as mandatory functionalities in case the FMC has to operate with 16 bits and 32 bits respectively.
> 
> The `periph_sdram` and `periph_nor_sram` features, on the other hand, could indeed be considered optional to control whether the SDRAM or the SRAM is used, e.g. when FMC is used for display but the SDRAM or the SRAM should not be used thinking

Hmm, ok. So maybe in the application Makefile use:

```mk
FEATURES_REQUIRED += periph_fmc
FEATURES_OPTIONAL += periph_fmc_sram
FEATURES_OPTIONAL += periph_fmc_sdram
```

and in the board Makefile.dep:

```mk
ifneq (,$(filter periph_fmc,$(USEMODULE)))
    FEATURES_REQUIRED += periph_fmc_16bit  # of 32bit depending on what the board is providing
endif
```

What do you think?

Reply:
> Hmm, ok. So maybe in the application Makefile use:
> 
> ```makefile
> FEATURES_REQUIRED += periph_fmc
> FEATURES_OPTIONAL += periph_fmc_sram
> FEATURES_OPTIONAL += periph_fmc_sdram
> ```
> 
> and in the board Makefile.dep:
> 
> ```makefile
> ifneq (,$(filter periph_fmc,$(USEMODULE)))
>     FEATURES_REQUIRED += periph_fmc_16bit  # or 32bit depending on what the board is providing
> endif
> ```
> 
> What do you think?

This is exactly like I understand the dependencies. I will try it in that way.

Reply:
Ok, I made the following changes:

- The test application is now in `tests/periph/fmc`.
- `periph_fmc_sdram` and `periph_fmc_nor_sram` are enabled by `FEATURES_OPTIONAL` in test application.
- `periph_fmc` is still enabled with `USEMODULE` in the test application to be consistent with `CONFIG_MODULE_PERIPH_FMC=y` which requires feature `periph_fmc` so that only boards with this feature can be compiled.
- `periph_fmc_16bit` is enabled by `FEATURE_REQUIRED` in board `Makefile.dep`.
- `MODULE_PERIPH_FMC_16BIT` and `MODULE_PERIPH_FMC_32BIT` are now hidden in Kconfig and enabled by `HAVE_PERIPH_FMC_16BIT` and `HAVE_PERIPH_FMC_32BIT`.
- RAM is enabled as heap in `tests/sys/malloc` if it is compiled with `USEMODULE=periph_fmc_sdram` or `USEMODULE=periph_fmc_nor_sram`. I'm not sure whether we should add FMC as `FEATURES_OPTIONAL` here.
- If SDRAM or NOR/PSRAM/SRAM is configured in board's `periph_conf.h` but the corresponding feature is not enabeled it is not initialized. It doesn't trigger an assertion anymore but only prints now a warning that the corresponding feature is not enabeld.

Reply:
> * I'm not sure whether we should add FMC as `FEATURES_OPTIONAL` here.

I don't know either. It could maybe increase the testing time because the heap can be much larger, so more message to print.



> * `periph_fmc` is still enabled with `USEMODULE` in the test application to be consistent with `CONFIG_MODULE_PERIPH_FMC=y` which requires feature `periph_fmc` so that only boards with this feature can be compiled.

In the Makefile, it should still be included using "FEATURES_REQUIRED += ". The features is turned into a `USEMODULE` automatically by the build system somewhere else. But it's important to use FEATURES_REQUIRED here so the build system is able to know what boards is supported (and it's also used by Murdock to skip the build on boards that doesn't provide this feature, I think).